### PR TITLE
Modmail reporting feature expansion

### PR DESCRIPTION
**Pull request by AzatothTheAwakened**
I worked on #13 with this branch.

Additional features that are not mentioned in the initial issue:
- archiving commands can have a note, the note can also be omitted

### CODE REVIEW

File: src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    public void handleModResponse(MessageReceivedEvent event) {
+
+        var modMessage = event.getMessage();
+        if (isInModMailChannel(event) && !event.getAuthor().isBot() && messageIsNotCommand(modMessage)) {
+            var participiant = getModMailParticipiant(event.getJDA(), getParticipiantId(event.getTextChannel()));
+            ModmailUtils.logInfo(LOGGER, String.format("[%s|%s] answered to case [%s|%s]", event.getAuthor().getAsTag(),
+                    event.getAuthor().getAvatarId(), participiant.getAsTag(), participiant.getId()));
+            participiant.openPrivateChannel().queue(channel -> {
+                channel.sendMessage("Staff reply:").queue(null, (ignore) -> modMessage.addReaction("❌").queue());
```
Comment (Toby-Larone) [inline src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java:28]: "Staff reply" text is not required

File: src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        return modmailChannel.getHistoryFromBeginning(1).complete().getRetrievedHistory().get(0).getEmbeds().stream()
+                .limit(1).map(embed -> embed.getFields()).flatMap(fields -> fields.stream())
+                .filter(field -> field.getName().equals("ID")).map(Field::getValue).map(Long::parseLong)
+                .reduce((result, ignore) -> result)
+                .orElseThrow(() -> new IllegalStateException(
+                        "Could not find embed with field 'ID' in the first message of channel: "
+                                + modmailChannel.getName()));
+    }
+
+    private Message formatDirectMessage(PrivateMessageReceivedEvent event) {
```
Comment (Toby-Larone) [inline src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java:68]: Can this method call just be replaced with `event.getMessage()`?

File: src/main/java/com/jvmrally/lambda/modmail/ModmailUtils.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * ModmailUtils
+ */
+public class ModmailUtils {
+
+    public static Optional<TextChannel> getReportsArchiveChannel(Guild guild) {
+        return guild.getTextChannelsByName("reports-archive", false).stream().reduce((x, ignore) -> x);
+    }
+
+    public static class Constants {
+        public static final String MODMAIL_COMMAND_PREFIX = "!modmail";
```
Comment (Toby-Larone) [inline src/main/java/com/jvmrally/lambda/modmail/ModmailUtils.java:20]: Should probably go in the modmail command class

File: src/main/java/com/jvmrally/lambda/modmail/ModmailUtils.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import net.dv8tion.jda.api.entities.TextChannel;
+
+import org.apache.logging.log4j.Logger;
+
+/**
+ * ModmailUtils
+ */
+public class ModmailUtils {
+
+    public static Optional<TextChannel> getReportsArchiveChannel(Guild guild) {
```
Comment (Toby-Larone) [inline src/main/java/com/jvmrally/lambda/modmail/ModmailUtils.java:15]: Could be a local method to the `ModmailChannelArchiver` class, or probably more preferably a more generic `getChannelByName` method

File: src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java (hunk 1 above)
Comment (AzatothTheAwakened) [reply src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java]: changed in:
9506fe5

File: src/main/java/com/jvmrally/lambda/modmail/ModmailUtils.java (hunk 3 above)
Comment (AzatothTheAwakened) [reply src/main/java/com/jvmrally/lambda/modmail/ModmailUtils.java]: changed in:
44f78e4

File: src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java (hunk 2 above)
Comment (AzatothTheAwakened) [reply src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java:66]: Which method call?

File: src/main/java/com/jvmrally/lambda/modmail/ModmailUtils.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        return guild.getTextChannelsByName("reports-archive", false).stream().reduce((x, ignore) -> x);
+    }
+
+    public static class Constants {
+        public static final String MODMAIL_COMMAND_PREFIX = "!modmail";
+    }
+
+    public static void logInfo(Logger logger, String message) {
```
[the remark this answers is not in the document]
Comment (AzatothTheAwakened) [reply src/main/java/com/jvmrally/lambda/modmail/ModmailUtils.java:19]: I am honest. I only used it to not duplicate the formatting all the time. Didn't even think of abstraction. Should I still change it?

File: src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    private User getModMailParticipiant(JDA jda, long participiantId) {
+        return jda.getGuilds().stream().map(guild -> guild.getMemberById(participiantId))
+                .reduce((result, ignore) -> result)
+                .orElseThrow(() -> new IllegalStateException("Could not find user with ID: " + participiantId))
+                .getUser();
+    }
+
+    private long getParticipiantId(TextChannel modmailChannel) {
+        return modmailChannel.getHistoryFromBeginning(1).complete().getRetrievedHistory().get(0).getEmbeds().stream()
```
[the remark this answers is not in the document]
Comment (AzatothTheAwakened) [reply src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java]: changed in:
13e1da2

File: src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java (hunk 2 above)
Comment (AzatothTheAwakened) [reply src/main/java/com/jvmrally/lambda/modmail/ModmailCommunicationHandler.java]: changed in:
5d8a3bf

**Pull request merged**